File: DesignPatternsNew/observerpattern-app/src/com/techlabs/state/Account.java
package com.techlabs.state;

import java.util.List;
import java.util.ArrayList;

import com.techlabs.observer.AlertType;

public class Account implements State {

	private int accountNumber;
	private String name;
	private double balance;
	List<AlertType> alertlist=new ArrayList<AlertType>();
	
	public Account(int accountNumber,String name,double balance)
	{
		this.accountNumber=accountNumber;
		this.name=name;
		this.balance=balance;
	}
		
	public int getAccountNumber() {
		return accountNumber;
	}

	public String getName() {
		return name;
	}

	public double getBalance() {
		return balance;
	}

	public void deposit(double amount)
	{
		balance=balance+amount;
		String message="Amount Deposited";
		update(message);
	}
	
	public void withdraw(double amount)
	{
		if(balance-amount >= 500 && amount > 0)
		{
			balance=balance-amount;
			String message="Amount withdrawn";
			update(message);
		}
	}

	@Override
	public void addAlertOption(AlertType alert) {
		alertlist.add(alert);

		
	}

	@Override
	public void deleteAlertOption(AlertType alert) {
		alertlist.remove(alert);
	}
	
	@Override
	public void update(String message)
	{
		for(AlertType alert: alertlist)
		{
			alert.updateBalance(this,message);
		}
	}
}

File: CSharpOOPS/accountchange/accountchange/Account.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace accountchange
{
    class Account
    {
        private double _balance;
        private string _name;
        private string _number;
        static private int acc_number=101;

        public Account(string name)
        {
            _name = name;
            _number = ""+acc_number++;
            _balance = 1000;
        }

        public double Balance
        {
            get
            {
                return _balance;
            }
        }
        public string Name
        {
            get
            {
                return _name;
            }
        }
        public string Number
        {
            get
            {
                return _number;
            }
        }

        public string deposit(double amount)
        {
            _balance += amount;
            string message = "Amount deposit " + amount;
            return message;
        }

        public string withdraw(double amount)
        {
            string message="";
            if (_balance - amount >= 500)
            {
                _balance -= _balance - amount;
                message = "Amount withdarwn" + amount;
                return message;
            }
            return message;
        }
    }
}
